proposed/logging-generator.md

```diff
@@ -0,0 +1,308 @@
+
+## Scenarios
+
+Heads up that `LoggingSample1` and `LoggingSample2` show existing approaches to logging, and the remaining will illustrate the samples using the source generator approach with `LoggerMessageAttribute`.
+
+## Background / Existing Approaches
+
+Currently there are two forms possible for doing logging, using `Microsoft.Extensions.Logging`:
+
+```c#
+public class LoggingSample1
+{
+    private ILogger _logger;
+
+    public LoggingSample1(ILogger logger)
+    {
+        _logger = logger;
+    }
+
+    public void LogMethod(string name)
+    {
+        _logger.LogInformation("Hello {name}", name);
+    }
+}
+```
+
+Here are some problems with the `LoggingSample1` sample using `LogInformation`, `LogWarning`, etc.: 
+1. We cannot provide event ID through these APIs
+2. Parameters passed are processed before LogLevel checks
+3. It requires parsing of message string on every use to find templates to substitute
+
+Because of these problems, the more efficient approach recommended today as best practices is to use [LoggerMessage.Define](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/logging/loggermessage?view=aspnetcore-5.0) APIs instead, illustrated below with `LoggingSample2`:
+
+```c#
+public class LoggingSample2
+{
+    private ILogger _logger;
+
+    public LoggingSample2(ILogger logger)
+    {
+        _logger = logger;
+    }
+
+    public void LogMethod(string name)
+    {
+        Log.LogName(_logger, name);
+    }
+
+    private static class Log
+    {
+        private static readonly Action<ILogger, string, Exception> _logName = LoggerMessage.Define<string>(LogLevel.Information, 0, @"Hello {name}");
+
+        public static void LogName(ILogger logger, string name)
+        {
+            _logName(logger, name, null!);
+        }
+    }
+}
+```
+
+An argument against `LoggerMessage.Define` APIs through community has been that it is hard to use, and there is a maintainance burden that hurts usability.
+
+## Requirements
+
+### Goals
+
+The number one concern raised with logging seems to be, in terms of usability, on finding a way to enforce strong typing on logging. In other words, it would be great to come up with a strong convention for all developers in a project to follow a specific logging-only template.
+
+### Non-Goals
+
+Also it's been raised that `LoggerMessage.Define` approach triggers boxing of data, and we wanted to understand if there are ways to improve this with a new and improved design.
+
+
+## Design
+
+`LoggerMessageAttribute` As shown in the examples, is used by the developer to trigger source generation:
+
+```c#
+namespace Microsoft.Extensions.Logging
+{
+    [AttributeUsage(AttributeTargets.Method)]
+    public sealed partial class LoggerMessageAttribute : Attribute
+    {
+        public LoggerMessageAttribute(int eventId, LogLevel level, string? message = null);
+        public LoggerMessageAttribute(int eventId, string? message = null);
```

Comment:
Just curious, why is the event ID required? Would it make sense to add a constructor without an event ID?

Reply:
It's required because they uniquely identify the event within a category.

Reply:
Sorry, I still don’t understand the requirement as the `LoggingSample1` doesn’t provide an event ID. From my perspective, this loses the uniqueness benefit but gains some ergonomics in return. If I were to use `LoggerMessage` I would likely try to use `0` as my event ID as I personally don’t mind losing uniqueness.

Reply:
A 0 event id is wrong though, the generator will flag it and give you a compile time error. You need to specify unique event ids for your log messages. They are a contract between log producer and log consumer. It's the public API equivalent for log messages.

Reply:
Is it possible you won't know your eventId at compile time? Perhaps you store it in a config?

Reply:
Yep, but then you won't be using this source generator because its likely your format string also comes from that configuration file. At that point, you need a runtime API to read from configuration and "register" log messages. That's what `LoggerMessage.Define` gives you.

Reply:
> They are a contract between log producer and log consumer. It's the public API equivalent for log messages.

Could you say more about that contract? I see that [`LoggerExtensions.Log`](https://github.com/dotnet/runtime/blob/8445e4eaa3b18bfdbed1eb1507293777c590d371/src/libraries/Microsoft.Extensions.Logging.Abstractions/src/LoggerExtensions.cs#L372-L375) accepts a nullable message and defaults the event ID to 0. Does that mean that `LogInformation` and the likes break that contract if you skip the event ID parameter?

A little bit of context. I worked on SharePoint where all log statements have a unique event ID automatically generated. These logs served a dual purpose: to debug problems in production, and as signals of events for monitoring and analysis by data scientists. Unfortunately, engineers would often break monitors or data reports by changing the some logging logic. It was hard to tell which logging statements were used just for logging, and which ones were also used as signal events.

I'm now on the nuget.org team and we avoid this problem by separating logs (for debugging) and metrics (for signals of events). We avoid regressions in our metrics through testing. If we have gaps in our metrics, we can do adhoc analysis using our logs (our logs include the original message template, which we use as a proxy to an event ID).

TLDR: I'm skeptical of event IDs in logs, but maybe there are scenarios I'm unaware of :)

Reply:
> I'm now on the nuget.org team and we avoid this problem by separating logs (for debugging) and metrics (for signals of events). We avoid regressions in our metrics through testing. If we have gaps in our metrics, we can do adhoc analysis using our logs (our logs include the original message template, which we use as a proxy to an event ID).

The contract between a log producer and log consumer is that an EventId uniquely identifies a log (and not the message template).

If in your use case, you're willing to make a even stronger contract that the message template uniquely identifies a log, then that's okay too (until you have a typo that you're afraid to fix😛) 

> TLDR: I'm skeptical of event IDs in logs. I personally prefer separating logs and metrics.

Nothing in this proposal is asking you to compute metric from logs. I wholly agree with you that it's a bad idea for a multitude of reasons.

Reply:
> Thanks for the explanation! Could you say more about that contract? I see that LoggerExtensions.Log accepts a nullable message and defaults the event ID to 0. Does that mean that LogInformation and the likes break that contract if you skip the event ID parameter?

I'd recommend library authors not use that method. I suspect this was added a long time ago to placate application developers to try and reduce the concept counts in logging, but this is from before my time 🤷

Reply:
I had a chat with @shirhatti offline. What do y'all think of:

1. Clarifying in the goals of the document that this spec is trying to enforce best practices on library authors
2. Allowing uses of the `LoggerMessage` attribute without defining an event ID, but, adding a Roslyn Analyzer that warns/errors to recommend best practices

Reply:
I think it should error by default and you can turn it off if you don't care about event ids.

Reply:
addressed this by adding an extra Q&A section for the source gen in the document.